### PR TITLE
writing aquifer cell and connection information to INIT file

### PR DESCRIPTION
**Pull request by GitPaean**

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
This PR is writing AQUIFERA and AQUIFERN information to INIT file. From my side, it is ready for review.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I have not added a test. Suggestions about how to add a test for this pr is also appreciated.

**Comment by GitPaean** [comment]
jenkins build this update_data please

File: src/opm/output/eclipse/WriteInit.cpp [hunk 1]
```diff
@@ -32,7 +32,6 @@
 #include <opm/output/eclipse/WriteRestartHelpers.hpp>
 
 #include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
-#include <opm/parser/eclipse/EclipseState/EndpointScaling.hpp>
```
Comment (bska) [inline src/opm/output/eclipse/WriteInit.cpp:35]: Is there a reason for removing this header?  We definitely use the [`EndpointScaling`](https://github.com/OPM/opm-common/blob/a1ee2f3b07b6b78d44d1a1b6b46257cebef7c8cb/opm/parser/eclipse/EclipseState/EndpointScaling.hpp#L28) class when collecting and emitting the [end-point scaling flags](https://github.com/OPM/opm-common/blob/a1ee2f3b07b6b78d44d1a1b6b46257cebef7c8cb/src/opm/output/eclipse/WriteInit.cpp#L261-L266) of the [`LOGIHEAD`](https://github.com/OPM/opm-common/blob/a1ee2f3b07b6b78d44d1a1b6b46257cebef7c8cb/src/opm/output/eclipse/WriteInit.cpp#L280) vector.

File: src/opm/output/eclipse/WriteInit.cpp [hunk 2]
```diff
@@ -559,6 +558,63 @@ namespace {
...
+        const auto& aquifer_cells = num_aquifers.allAquiferCells();
+        for ([[maybe_unused]] const auto& [cell_idx, cell] : aquifer_cells) {
+            const size_t active_index = grid.activeIndex(cell->global_index);
+            aquifern[active_index] -= std::exp2(cell->aquifer_id - 1);
+        }
+
+        // aquifer connections
+        for (const auto& [id, aqu] : num_aquifers.aquifers()) {
+            const auto& connections = aqu.connections();
+            const int exp2_id_1 = std::exp2(id - 1);
```
Comment (bska) [inline src/opm/output/eclipse/WriteInit.cpp:578]: Note that `std::exp2()` returns a floating-point result.  Is there a reason for not using bitshifts here?
```C++
const int exp2_id_1 = 1 << (id - 1);
```

File: src/opm/output/eclipse/WriteInit.cpp [hunk 3]
```diff
@@ -559,6 +558,63 @@ namespace {
...
+            for (const auto& con : connections) {
+                const size_t active_index = grid.activeIndex(con.global_index);
+                aquifern[active_index] += exp2_id_1;
```
Comment (bska) [inline src/opm/output/eclipse/WriteInit.cpp:581]: Do we need to cater to the possibility that a single cell is connected to multiple aquifers?  If not, I'd prefer to just use regular assignment here.

**Review by bska — COMMENTED** [review]
Looks good for the most part.  I think it's more expressive to use integer operations for the powers of two and I think you remove a header that's actually used but otherwise I'm happy with this.

File: src/opm/output/eclipse/WriteInit.cpp (hunk 2 above)
Comment (GitPaean) [reply src/opm/output/eclipse/WriteInit.cpp:578]: I seldom use bitwise operation, so it was not in my mind. I will change to it. 

And also I knew `std::exp2()` return floating-point results, while I think it is pretty safe for this case.

File: src/opm/output/eclipse/WriteInit.cpp (hunk 3 above)
Comment (GitPaean) [reply src/opm/output/eclipse/WriteInit.cpp:581]: > Do we need to cater to the possibility that a single cell is connected to multiple aquifers? 

Yes, it looks like a single cell can be connected to multiple aquifers. If other implementation does not allow the aquifers to do so, we need to correct them. But here, let us keep the support.

File: src/opm/output/eclipse/WriteInit.cpp (hunk 1 above)
Comment (GitPaean) [reply src/opm/output/eclipse/WriteInit.cpp:35]: You are right.  It is the IDE I am using show this header file is not used, and I went through quickly, did not find it used, which was a mistake.

FYI, and I just checked the standard from the IDE, it is based on the following standard to mark this header file not used. 
```
Detect not directly used
an include directive is unused if declarations from it are not used in the file directly. This works best with the "Include What You Use" policy.
```

File: src/opm/output/eclipse/WriteInit.cpp [hunk 4]
```diff
@@ -559,6 +558,63 @@ namespace {
...
+    void writeNumericalAquifers(const Opm::NumericalAquifers& num_aquifers,
+                                const ::Opm::EclipseGrid&          grid,
+                                ::Opm::EclIO::OutputStream::Init&  initFile)
+    {
+        std::vector<int> aquifern(grid.getNumActive(), 0);
+        // aquifer cells
+        const auto& aquifer_cells = num_aquifers.allAquiferCells();
+        for ([[maybe_unused]] const auto& [cell_idx, cell] : aquifer_cells) {
+            const size_t active_index = grid.activeIndex(cell->global_index);
+            aquifern[active_index] -= std::exp2(cell->aquifer_id - 1);
```
Comment (GitPaean) [inline src/opm/output/eclipse/WriteInit.cpp:572]: Different from the connections, an aquifer cell should not belong to two aquifers. Here should use the assignment. Although it should not change things in the results.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=484 please

**Comment by bska** [comment]
Thanks for the updates.  This looks good to me now so I'll merge this now.

**Pull request merged**